sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -444,12 +444,22 @@ object ColumnPruning extends Rule[LogicalPlan] {
       f.copy(child = prunedChild(child, f.references))
     case e @ Expand(_, _, child) if (child.outputSet -- e.references).nonEmpty =>
       e.copy(child = prunedChild(child, e.references))
-    case g: Generate if !g.join && (g.child.outputSet -- g.references).nonEmpty =>
-      g.copy(child = prunedChild(g.child, g.references))
 
-    // Turn off `join` for Generate if no column from it's child is used
-    case p @ Project(_, g: Generate) if g.join && p.references.subsetOf(g.generatedSet) =>
-      p.copy(child = g.copy(join = false))
+    // prune unrequired references
+    case g : Generate
+      if (AttributeSet(g.unrequiredChildOutput) -- g.generator.references).nonEmpty =>
+        g.copy(child = prunedChild(g.child,
+                        g.child.outputSet -- g.unrequiredChildOutput ++ g.generator.references),
+               unrequiredChildOutput = g.generator.references.toSeq)
+
+    // Sync Generate's unrequiredChildOutput with the actual needed outputs
+    case p @ Project(_, g: Generate) =>
```

Comment:
seems like we can merge this case with the above case, as column pruning only works if there is a `Project` above `Generate`
```
case p @ Project(_, g: Generate) =>
  if (p.preference == g.outputSet) {
    p
  } else {
    val requiredAttrs = p.reference -- g.producedAttributes ++ g.generator.reference
    val newChild = pruneChild(g.child, requiredAttrs)
    p.copy(child = g.copy(child = newChild, unrequiredChildOutput = g.generator.reference -- p.reference))
  }
```

Reply:
Yep, I agree.
although, I'm not sure why do we "swallow" the this case up to here and not letting it proceed to other rules like:
    // Eliminate no-op Projects
    case p @ Project(_, child) if sameOutput(child.output, p.output) => child

because in effect the Generate is now doing the job of the Project

Reply:
makes sense, we should move this rule below `case p @ Project(_, child) if sameOutput(child.output, p.output) => child`